spec/controllers/local_authorities_controller_spec.rb

```diff
@@ -63,18 +63,73 @@
+    context "with a valid CSV" do
+      before { @local_authority = create(:local_authority) }
+      let(:path) { Rails.root.join("spec/lib/local-links-manager/import/fixtures/imported_links.csv") }
+      let(:csv) { Rack::Test::UploadedFile.new(path, "text/csv", true) }
+      let(:url_regex) { /http:\/\/.+\/local_authorities\/#{@local_authority.slug}/ }
 
-    it "retrieves HTTP found" do
-      login_as_stub_user
-      post(:upload_links_csv, params: { local_authority_slug: @local_authority.slug, csv: })
+      it "retrieves HTTP found" do
+        login_as_stub_user
+        post(:upload_links_csv, params: { local_authority_slug: @local_authority.slug, csv: })
+
+        expect(response.status).to eq(302)
+        expect(response.location).to match(url_regex)
+        expect(response.headers["Content-Type"]).to eq("text/html; charset=utf-8")
+      end
+    end
 
-      expect(response.status).to eq(302)
-      expect(response.location).to match(url_regex)
-      expect(response.headers["Content-Type"]).to eq("text/html; charset=utf-8")
+    context "with errors in the CSV" do
+      before do
+        @local_authority = create(:local_authority)
+        service1 = create(:service, lgsl_code: 1)
+        service2 = create(:service, lgsl_code: 2)
+        service3 = create(:service, lgsl_code: 3)
+        service4 = create(:service, lgsl_code: 4)
+        service5 = create(:service, lgsl_code: 5)
+        service6 = create(:service, lgsl_code: 6)
+        interaction = create(:interaction, lgil_code: 1)
+        create(:service_interaction, service: service1, interaction:)
+        create(:service_interaction, service: service2, interaction:)
+        create(:service_interaction, service: service3, interaction:)
+        create(:service_interaction, service: service4, interaction:)
+        create(:service_interaction, service: service5, interaction:)
+        create(:service_interaction, service: service6, interaction:)
+      end
+
+      let(:path_all) { Rails.root.join("spec/lib/local-links-manager/import/fixtures/imported_links_all_errors.csv") }
+      let(:csv_all) { Rack::Test::UploadedFile.new(path_all, "text/csv", true) }
+      let(:path_many) { Rails.root.join("spec/lib/local-links-manager/import/fixtures/imported_links_many_errors.csv") }
+      let(:csv_many) { Rack::Test::UploadedFile.new(path_many, "text/csv", true) }
+      let(:path_few) { Rails.root.join("spec/lib/local-links-manager/import/fixtures/imported_links_few_errors.csv") }
+      let(:csv_few) { Rack::Test::UploadedFile.new(path_few, "text/csv", true) }
+      let(:path_nothing) { Rails.root.join("spec/lib/local-links-manager/import/fixtures/imported_links_nothing_to_import.csv") }
+      let(:csv_nothing) { Rack::Test::UploadedFile.new(path_nothing, "text/csv", true) }
+      let(:url_regex) { /http:\/\/.+\/local_authorities\/#{@local_authority.slug}/ }
+
+      it "shows the all error message if all lines are broken" do
+        login_as_stub_user
```

Comment:
We could add `login_as_stub_user` to the `before` block rather than include in each test